server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSDocumentOptionsCheckUtility.java

```diff
@@ -0,0 +1,153 @@
+    }
+
+    /**
+     * Entrypoint to check CICS DOCUMENT rule options
+     *
+     * @param ctx ParserRuleContext subclass containing options
+     * @param <E> A subclass of ParserRuleContext
+     */
+    public <E extends ParserRuleContext> void checkOptions(E ctx) {
+        switch (ctx.getRuleIndex()) {
+            case CICSParser.RULE_cics_document_create:
+                checkDocumentCreate((CICSParser.Cics_document_createContext) ctx);
+                break;
+            case CICSParser.RULE_cics_document_delete:
+                checkDocumentDelete((CICSParser.Cics_document_deleteContext) ctx);
+                break;
+            case CICSParser.RULE_cics_document_insert:
+                checkDocumentInsert((CICSParser.Cics_document_insertContext) ctx);
+                break;
+            case CICSParser.RULE_cics_document_retrieve:
+                checkDocumentRetrieve((CICSParser.Cics_document_retrieveContext) ctx);
+                break;
+            case CICSParser.RULE_cics_document_set:
+                checkDocumentSet((CICSParser.Cics_document_setContext) ctx);
+                break;
+            default:
+                break;
+        }
+        checkDuplicates(ctx);
+    }
+
+    @SuppressWarnings("unchecked")
+    private void checkDocumentCreate(CICSParser.Cics_document_createContext ctx) {
+        checkHasMandatoryOptions(ctx.CREATE(), ctx, "CREATE");
+        checkHasMandatoryOptions(ctx.DOCTOKEN(), ctx, "DOCTOKEN");
+        if (!ctx.LENGTH().isEmpty()) {
+            checkHasExactlyOneOption("FROM, TEXT or BINARY", ctx, ctx.FROM(), ctx.TEXT(), ctx.BINARY());
+        }
+        checkHasMutuallyExclusiveOptions("LENGTH, FROMDOC or TEMPLATE",
+                ctx.LENGTH(), ctx.FROMDOC(), ctx.TEMPLATE());
+        if (!ctx.DELIMITER().isEmpty() || !ctx.UNESCAPED().isEmpty()) {
+            checkHasMandatoryOptions(ctx.SYMBOLLIST(), ctx, "SYMBOLLIST");
+            checkHasMandatoryOptions(ctx.LISTLENGTH(), ctx, "LISTLENGTH");
+        }
+        checkAllOptionsArePresentOrAbsent("SYMBOLLIST, LISTLENGTH", ctx, ctx.SYMBOLLIST(), ctx.LISTLENGTH());
+    }
+
+    private void checkDocumentDelete(CICSParser.Cics_document_deleteContext ctx) {
+        checkHasMandatoryOptions(ctx.DELETE(), ctx, "DELETE");
+        checkHasMandatoryOptions(ctx.DOCTOKEN(), ctx, "DOCTOKEN");
+    }
+
+    @SuppressWarnings("unchecked")
+    private void checkDocumentInsert(CICSParser.Cics_document_insertContext ctx) {
+        checkHasMandatoryOptions(ctx.INSERT(), ctx, "INSERT");
+        checkHasMandatoryOptions(ctx.DOCTOKEN(), ctx, "DOCTOKEN");
+        if (!ctx.LENGTH().isEmpty()) {
+            checkHasExactlyOneOption("FROM, TEXT or BINARY", ctx, ctx.FROM(), ctx.TEXT(), ctx.BINARY());
+        }
+        checkHasExactlyOneOption("LENGTH, SYMBOL, TEMPLATE, FROMDOC or BOOKMARK", ctx,
+                ctx.LENGTH(), ctx.SYMBOL(), ctx.TEMPLATE(), ctx.FROMDOC(), ctx.BOOKMARK());
```

Comment:
I think it would be better to include FROM, TEXT and BINARY instead of LENGTH.
And then required LENGTH for them.